File: python/src/problems1_100/string_to_integer_atoi.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def my_atoi(self, s: str) -> int:
        pass


class Default(Problem):
    @override
    def my_atoi(self, s: str) -> int:
        if not s:
            return 0
        s = s.strip()
        negate = False
        if s[0] == '-':
            s = s[1:]
            negate = True
        elif s[0] == '+':
            s = s[1:]

        result = ''
        for c in s:
            if c.isdigit():
                result += c
                continue
            break
        if negate:
            result = '-' + result
        if len(result) == 0 or \
            (len(result) == 1 and (result[0] == '-' or result[0] == '+')):
            return 0
        return int(result)


SOLUTIONS: list[Problem] = [Default()]

File: javascript/src/problems1_100/string-to-integer-atoi.js
class Problem {
  myAtoi(s) {
    throw new Error(`'Not implemented (${s}).`);
  }
}

class Default extends Problem {
  myAtoi(s) {
    if (!s) {
      return 0;
    }
    s = s.trim();
    let negate = false;
    if (s[0] === '-') {
      s = s.slice(1);
      negate = true;
    } else if (s[0] === '+') {
      s = s.slice(1);
    }

    let result = '';
    for (let c of s) {
      if (c >= '0' && c <= '9') {
        result += c;
        continue;
      }
      break;
    }
    if (negate) {
      result = `-${result}`;
    }
    if (result.length === 0 ||
        (result.length === 1 && (result[0] === '-' || result[0] === '+'))) {
      return 0;
    }
    return parseInt(result);
  }
}

export const SOLUTIONS = [new Default()];
